File: packages/lmon/src/stringifier.ts
import { StringifyOptions } from './ast';

export function stringify(
  input: Record<string, unknown> | unknown[],
  options: StringifyOptions = {},
): string {
  const {
    rowLabels = true,
  } = options;

  if (Array.isArray(input)) {
    return stringifyArray(input, rowLabels);
  }

  return stringifyObject(input, rowLabels);
}

function stringifyArray(arr: unknown[], rowLabels: boolean): string {
  if (arr.length === 0) {
    return '';
  }

  const first = arr[0];

  // Infer schema from first element
  if (typeof first === 'object' && first !== null && !Array.isArray(first)) {
    const schema = inferSchemaFromObject(first as Record<string, unknown>);
    const headerLine = formatHeader(schema);

    const rows = arr
      .map((item) => {
        const docStr = formatDocument(item as Record<string, unknown>, schema);
        return docStr;
      })
      .join('\n');

    return headerLine ? `${headerLine}\n${rows}` : rows;
  }

  // Array of scalars
  const docStr = formatDocument(arr, []);
  return docStr;
}

function stringifyObject(
  obj: Record<string, unknown>,
  rowLabels: boolean,
): string {
  if (Object.keys(obj).length === 0) {
    return '';
  }

  const firstValue = Object.values(obj)[0];

  // Infer schema from first value
  if (typeof firstValue === 'object' && firstValue !== null && !Array.isArray(firstValue)) {
    const schema = inferSchemaFromObject(firstValue as Record<string, unknown>);
    const headerLine = formatHeader(schema);

    const rows = Object.entries(obj)
      .map(([key, value]) => {
        const docStr = formatDocument(value as Record<string, unknown>, schema);
        return rowLabels ? `${key}:${docStr}` : docStr;
      })
      .join('\n');

    return headerLine ? `${headerLine}\n${rows}` : rows;
  }

  // Simple object with scalar values
  const schema = inferSchemaFromObject(obj);
  const headerLine = formatHeader(schema);

  const rows = Object.entries(obj)
    .map(([key, value]) => {
      const docStr = formatDocument(value as Record<string, unknown>, schema);
      return rowLabels ? `${key}:${docStr}` : docStr;
    })
    .join('\n');

  return headerLine ? `${headerLine}\n${rows}` : rows;
}

function inferSchemaFromObject(
  obj: unknown,
): { name: string; isArray: boolean; children: { name: string; isArray: boolean; children: unknown[] }[] }[] {
  if (!obj || typeof obj !== 'object') return [];

  if (Array.isArray(obj)) return [];

  const entries = Object.entries(obj as Record<string, unknown>);
  return entries.map(([key, value]) => {
    const isArray = Array.isArray(value);
    const children =
      typeof value === 'object' && value !== null && !Array.isArray(value)
        ? inferSchemaFromObject(value)
        : [];

    return {
      name: key,
      isArray,
      children,
    };
  });
}

function formatHeader(
  schema: { name: string; isArray: boolean; children: unknown[] }[],
): string {
  if (schema.length === 0) return '';

  const labels = schema.map((s) => {
    const arrayMarker = s.isArray ? '[]' : '';
    if (s.children.length > 0) {
      const nestedLabels = (s.children as { name: string }[])
        .map((c) => c.name)
        .join(',');
      return `${s.name}:(${nestedLabels})`;
    }
    return `${s.name}${arrayMarker}`;
  });

  return `(${labels.join(',')})`;
}

function formatDocument(
  value: unknown,
  _schema: unknown[],
): string {
  if (Array.isArray(value)) {
    const items = value.map((v) => formatValue(v));
    return `{${items.join(',')}}`;
  }

  if (typeof value === 'object' && value !== null) {
    const entries = Object.entries(value as Record<string, unknown>);
    const items = entries.map(([, v]) => formatValue(v));
    return `{${items.join(',')}}`;
  }

  return `{${formatValue(value)}}`;
}

function formatValue(value: unknown): string {
  if (value === null) {
    return 'null';
  }

  if (typeof value === 'boolean') {
    return value ? 'true' : 'false';
  }

  if (typeof value === 'number') {
    return String(value);
  }

  if (Array.isArray(value)) {
    const items = value.map((v) => formatValue(v));
    return `[${items.join(',')}]`;
  }

  if (typeof value === 'object' && value !== null) {
    const entries = Object.entries(value as Record<string, unknown>);
    const items = entries.map(([, v]) => formatValue(v));
    return `{${items.join(',')}}`;
  }

  // String - may need escaping
  const str = String(value);
  if (needsQuoting(str)) {
    return `"${escapeString(str)}"`;
  }
  return str;
}

function needsQuoting(str: string): boolean {
  if (str === '') return false;
  // These chars require quoting, not escaping
  if (str.includes(' ')) return true;
  if (str.includes('"')) return true;
  if (str.includes("'")) return true;
  // Everything else can be escaped or quoted - we choose quoting for simplicity
  if (/[,{}[\]:]/.test(str)) return true;
  return false;
}

function escapeString(str: string): string {
  return str
    .replace(/\\/g, '\\\\')
    .replace(/"/g, '\\"')
    .replace(/\n/g, '\\n')
    .replace(/\t/g, '\\t');
}

File: packages/xcon-python/xcon/stringifier.py
from typing import Any, Dict, List, Optional


def stringify(obj: Any) -> str:
    """Convert Python object to XCON string."""
    if isinstance(obj, list):
        return stringify_array(obj)
    elif isinstance(obj, dict):
        return stringify_object(obj)
    else:
        return stringify_array([obj])


def stringify_array(arr: list) -> str:
    """Stringify a list."""
    if not arr:
        return ""

    first = arr[0] if arr else None

    # Infer schema from first element
    if isinstance(first, dict):
        schema = infer_schema_from_object(first)
        header_line = format_header(schema)

        rows = "\n".join(
            format_document(item, schema) for item in arr
        )

        return f"{header_line}\n{rows}" if header_line else rows

    # Array of arrays (no header)
    if isinstance(first, list):
        rows = "\n".join(
            format_document(item, []) for item in arr
        )
        return rows

    # Array of scalars
    doc_str = format_document(arr, [])
    return doc_str


def stringify_object(obj: Dict[str, Any]) -> str:
    """Stringify a dict."""
    if not obj:
        return ""

    first_value = next(iter(obj.values())) if obj else None

    # Infer schema from first value
    if isinstance(first_value, dict):
        schema = infer_schema_from_object(first_value)
        header_line = format_header(schema)

        rows = "\n".join(
            f"{key}:{format_document(value, schema)}"
            for key, value in obj.items()
        )

        return f"{header_line}\n{rows}" if header_line else rows

    # Simple object with scalar values
    schema = infer_schema_from_object(obj)
    header_line = format_header(schema)

    rows = "\n".join(
        f"{key}:{format_document(value, schema)}"
        for key, value in obj.items()
    )

    return f"{header_line}\n{rows}" if header_line else rows


def infer_schema_from_object(obj: Any) -> List[Dict[str, Any]]:
    """Infer schema from an object."""
    if not isinstance(obj, dict):
        return []

    schema = []
    for key, value in obj.items():
        is_array = isinstance(value, list)
        children = (
            infer_schema_from_object(value)
            if isinstance(value, dict)
            else []
        )

        schema.append({
            'name': key,
            'is_array': is_array,
            'children': children,
        })

    return schema


def format_header(schema: List[Dict[str, Any]]) -> str:
    """Format schema as XCON header."""
    if not schema:
        return ""

    labels = []
    for s in schema:
        array_marker = "[]" if s['is_array'] else ""
        if s['children']:
            nested_labels = ",".join(c['name'] for c in s['children'])
            labels.append(f"{s['name']}:({nested_labels})")
        else:
            labels.append(f"{s['name']}{array_marker}")

    return f"({','.join(labels)})"


def format_document(value: Any, schema: List[Dict[str, Any]]) -> str:
    """Format a value as a XCON document."""
    if isinstance(value, list):
        items = [format_value(v) for v in value]
        return "{" + ",".join(items) + "}"

    if isinstance(value, dict):
        items = [format_value(v) for v in value.values()]
        return "{" + ",".join(items) + "}"

    return "{" + format_value(value) + "}"


def format_value(value: Any) -> str:
    """Format a value."""
    if value is None:
        return "null"

    if isinstance(value, bool):
        return "true" if value else "false"

    if isinstance(value, (int, float)):
        return str(value)

    if isinstance(value, list):
        items = [format_value(v) for v in value]
        return "[" + ",".join(items) + "]"

    if isinstance(value, dict):
        items = [format_value(v) for v in value.values()]
        return "{" + ",".join(items) + "}"

    # String - may need quoting
    s = str(value)
    if needs_quoting(s):
        return '"' + escape_string(s) + '"'
    return s


def needs_quoting(s: str) -> bool:
    """Check if a string needs quoting."""
    if s == "":
        return False
    if " " in s:
        return True
    if '"' in s or "'" in s:
        return True
    if any(c in s for c in ",{}[]:\\"):
        return True
    return False


def escape_string(s: str) -> str:
    """Escape special characters in a string."""
    s = s.replace("\\", "\\\\")
    s = s.replace('"', '\\"')
    s = s.replace("\n", "\\n")
    s = s.replace("\t", "\\t")
    return s
